### DISCUSSION TITLE
TypeScript return types are inaccurate

### DISCUSSION THREAD

**Discussion by DKeen0123** [prisma/prisma#13330] [Q&A]
Hi all!

I'm having an issue with my Typescript types and looking to see if anyone can help me out with it.

Firstly, this issue is only happening because I want to create reusable getter and setter functions that wraps prisma stuff. 

So, an example:

I have a `User` model that can have many `Projects` and can have one `Country`. 

I have a wrapper function called `getUser` that looks like this:

![image](https://user-images.githubusercontent.com/16704395/168113684-65a69a7f-a93b-48be-a411-36bbf6aecb77.png)

The problem here is that the inferred return type looks like this:

![image](https://user-images.githubusercontent.com/16704395/168113979-eaccfbc3-8e78-4532-ad65-0a93358390bc.png)

now I would have expected that the `Projects` and `Country` on the return type would be optional or possibly null, however  they are not. This causes issues when calling this function as I can not pass in any options to this function, but my code will still say that the type is an expanded user with these two relations, which of course leads to broken code.

Does anyone know of a good way to make this function more intelligent?

**Comment by DKeen0123** [comment]
## If anyone has a cleaner way to do this, please do let me know!

I found a [related thread](https://github.com/prisma/prisma/discussions/2360) and tried out function overloading. It seems to work for my usecase! 

Check it out: 

I first need to define a couple of functions to get nicer types from prisma:

```typescript
export const getUserWithRelations = async ({
  isExternalId,
  id,
}: {
  isExternalId?: boolean;
  id: string;
}) => {
  let where: Prisma.UserWhereInput = {
    id,
  };
  if (isExternalId) {
    where = {
      externalAccountId: id,
    };
  }
  return await prismadb.user.findFirst({
    where,
    include: {
      Projects: true,
      Country: true,
      Businesses: true,
    },
  });
};
export const getUserWithOutRelations = async ({
  isExternalId,
  id,
}: {
  isExternalId?: boolean;
  id: string;
}) => {
  let where: Prisma.UserWhereInput = {
    id,
  };
  if (isExternalId) {
    where = {
      externalAccountId: id,
    };
  }
  return await prismadb.user.findFirst({
    where,
  });
};

type UserWithRelations = Prisma.PromiseReturnType<typeof getUserWithRelations>;
type UserWithoutRelations = Prisma.PromiseReturnType<
  typeof getUserWithOutRelations
>;
```
I define a `getUser` function with overload functions. This way, the return will be either just the user or the user with all its relations included, based on the inputted arguments: 
 
 ```typescript
export async function getUser(id: string, withRelations: true): Promise<UserWithRelations>;
export async function getUser(id: string): Promise<UserWithoutRelations>;
export async function getUser(id: string, withRelations?: boolean) {
  if (withRelations) {
    return await getUserWithRelations({
      id,
      isExternalId: false,
    });
  }

  return await getUserWithOutRelations({
    id,
    isExternalId: false,
  });
}

```

So now, if I were to call `getUser` with just an id, it would give me the type back as just `User` without relations!

It would be nice to be able to pass in the relations that I want to include on each individual call of the function. I might see if I can do that with some more overload functions.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
